### PR TITLE
librbd: discard hangs when 'rbd_skip_partial_discard' is enabled

### PR DESCRIPTION
**Pull request by trociny**
Fixes: http://tracker.ceph.com/issues/16386

### CODE REVIEW

**Comment by trociny** [comment]
@dillaman It fixes hang but I observe this test failure:
```
[ RUN      ] TestInternal.DiscardCopyup
test/librbd/test_internal.cc:682: Failure
Value of: bl.contents_equal(read_bl)
  Actual: false
Expected: true
[  FAILED  ] TestInternal.DiscardCopyup (24 ms)
```
Am I interpreting the test result correctly, that discard became visible for the snapshot?

**Comment by dillaman** [comment]
@trociny On line 676, I hide the fact that this clone has a parent to ensure it doesn't satisfy the read from the parent. If ```rbd_skip_partial_discard``` is enabled, nothing will be updated if it is a partial object discard, so this test should probably be skipped if it's enabled.

**Comment by dillaman** [comment]
lgtm

**Comment by dillaman** [comment]
@trociny I think we should create a ceph-qa-suite test case where we enable this option and run ceph_test_librbd

**Comment by dillaman** [comment]
@trociny ... also, I wonder if we should record this config setting in the discard journal event. This seems like it is the only config option we have that could result in a mismatched image between primary and non-primary should the settings be different between clusters.

**Comment by trociny** [comment]
@dillaman I was observing this failure:
```
[ RUN      ] TestImageSync.Simple
librbd/AioCompletion.cc: In function 'void librbd::AioCompletion::associate_journal_event(uint64_t)' thread 7fef1d70b700 time 2016-07-01 09:24:34.881967
librbd/AioCompletion.cc: 201: FAILED assert(state == STATE_PENDING)
 ceph version 11.0.0-122-gc8919c2 (c8919c2af59a924cf5202894a10afd0ec7ffe6d6)
 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x82) [0x7fef31fa58f2]
 2: (librbd::AioCompletion::associate_journal_event(unsigned long)+0x51) [0x7fef31d937f1]
 3: (librbd::AioImageDiscard::append_journal_event(std::list<librbd::AioObjectRequest*, std::allocator<librbd::AioObjectRequest*> > const&, bool)+0x89) [0x7fef31e82939]
 4: (librbd::AbstractAioImageWrite::send_request()+0x495) [0x7fef31e81e85]
 5: (librbd::AioImageRequest<librbd::ImageCtx>::send()+0xc4) [0x7fef31e83334]
 6: (librbd::AioImageRequestWQ::process(librbd::AioImageRequest<librbd::ImageCtx>*)+0x80) [0x7fef31d97490]
 7: (ThreadPool::worker(ThreadPool::WorkThread*)+0xb7f) [0x7fef31f9652f]
 8: (ThreadPool::WorkThread::entry()+0x10) [0x7fef31f97580]
 9: (()+0x80a4) [0x7fef28ab40a4]
 10: (clone()+0x6d) [0x7fef26df904d]
```
So I have made associate_journal_event() call in AioImageDiscard::append_journal_event() conditional.

May be a more cleaner solution is to add a new AioObjectRequest class to handle rbd_skip_partial_discard (see [DNM] commit)?

Both ways fixes the assert, but this failure is observed:
```
[ RUN      ] TestMockImageSyncObjectCopyRequest.Trim
seed 2386
 wrote [48644~207480]
unknown file: Failure

Unexpected mock function call - returning directly.
    Function call: aio_update(3, 0, 1, '\x1' (1), @0x7f5791ff87b0 --, 0x7f577c000b90)
Google Mock tried the following 2 expectations, but none matched:

test/rbd_mirror/image_sync/test_mock_ObjectCopyRequest.cc:175: tried expectation #0: EXPECT_CALL(mock_object_map, aio_update(snap_id, 0, 1, state, _, _))...
         Expected: all pre-requisites are satisfied
           Actual: the following immediate pre-requisites are not satisfied:
test/rbd_mirror/image_sync/test_mock_ObjectCopyRequest.cc:153: pre-requisite #0
                   (end of pre-requisites)
         Expected: to be called once
           Actual: never called - unsatisfied and active
test/rbd_mirror/image_sync/test_mock_ObjectCopyRequest.cc:175: tried expectation #1: EXPECT_CALL(mock_object_map, aio_update(snap_id, 0, 1, state, _, _))...
  Expected arg #0: is equal to 4
           Actual: 3
         Expected: to be called once
           Actual: never called - unsatisfied and active
```

**Comment by trociny** [comment]
@dillaman I will create  a ceph-qa-suite test for rbd_skip_partial_discard option.

**Comment by trociny** [comment]
> I wonder if we should record this config setting in the discard journal event. This seems like it is the only config option we have that could result in a mismatched image between primary and non-primary should the settings be different between clusters.

I don't have a strong opinion here. For me it might be not necessary and may be not expected behavior for some users. Also taking into account that we plan to replicate metadata options, if this is enabled/disabled per image, we will replicate this.

I don't have an idea how useful (frequently used) this option in real life to bother with this.

**Comment by dillaman** [comment]
@trociny I opened a new ticket to track it as a feature.  It doesn't matter now, but I can definitely see the potential for the future "rbd mirror image verify" command to flag deltas in the image.

**Comment by dillaman** [comment]
@trociny You could introduce a new virtual method ```prune_object_extents``` that, when skip partial discards is enabled, it will yank those corresponding extents.  You would call that before ```AioCompletion::set_request_count```

**Comment by trociny** [comment]
@dillaman Remade using prune_object_extents.

In TestMockImageSyncObjectCopyRequest.Trim for discard to truncate the test object I force setting rbd_skip_partial_discard to false for remote image. Does it looks ok for you?

**Comment by trociny** [comment]
@dillaman ceph-qa tests with rbd_skip_partial_discard option enabled: https://github.com/ceph/ceph-qa-suite/pull/

**Comment by dillaman** [comment]
lgtm

File: src/test/librbd/journal/test_Entries.cc [hunk 1]
```diff
@@ -172,6 +172,12 @@ TEST_F(TestJournalEntries, AioDiscard) {
   ASSERT_EQ(0, c->wait_for_complete());
   c->put();
 
+  CephContext* cct = reinterpret_cast<CephContext*>(_rados.cct());
+  if (cct->_conf->rbd_skip_partial_discard) {
+    ASSERT_FALSE(wait_for_entries_available(ictx));
```
Comment (dillaman) [inline src/test/librbd/journal/test_Entries.cc]: The discard journal entry should still be created -- why was this needed?

Comment (trociny) [inline src/test/librbd/journal/test_Entries.cc]: I did not see journal entry being created and I thought it was ok.

I suppose for journal entry to be created, `associate_journal_event()` should be called, which is in `AioImageDiscard::append_journal_event()`, and this only happens when sending object extents?

Comment (trociny) [inline src/test/librbd/journal/test_Entries.cc]: Actually, recalling you created a ticket to track rbd_skip_partial_discard in journal entries, I was going to discuss this observation but forgot.

Comment (trociny) [inline src/test/librbd/journal/test_Entries.cc]: @dillaman updated (skipping the test if partial discards are enabled).

**Comment by dillaman** [comment]
@trociny looks like the python tests need to be updated:

http://qa-proxy.ceph.com/teuthology/jdillaman-2016-07-22_04:36:50-rbd-wip-jd-testing---basic-vps/328662/teuthology.log

**Comment by trociny** [comment]
@dillaman I have updated the tests (since last your review and testing the changes are only in fsx.cc and test_rbd.py)

I run some teuthology tests and no issues are noticed. Tests you reported failed now pass:
```teuthology-suite --suite rbd --ceph wip-mgolub-testing --owner mgolub --email mgolub@mirantis.com --machine-type vps --filter 'rbd/librbd/{cache/none.yaml cachepool/none.yaml clusters/{fixed-3.yaml openstack.yaml} config/skip-partial-discard.yaml fs/xfs.yaml msgr-failures/few.yaml workloads/python_api_tests_with_journaling.yaml}'
http://pulpito.ceph.com/trociny-2016-07-28_01:57:05-rbd-wip-mgolub-testing---basic-vps/

teuthology-suite --suite rbd --ceph wip-mgolub-testing --owner mgolub --email mgolub@mirantis.com --machine-type vps --filter 'rbd/librbd/{cache/none.yaml cachepool/small.yaml clusters/{fixed-3.yaml openstack.yaml} config/skip-partial-discard.yaml fs/xfs.yaml msgr-failures/few.yaml workloads/fsx.yaml}'
http://pulpito.ceph.com/trociny-2016-07-28_01:57:46-rbd-wip-mgolub-testing---basic-vps/
```
Also, I have been running the following tests:
```
teuthology-suite --suite rbd --ceph wip-mgolub-testing --owner mgolub --email mgolub@mirantis.com --machine-type vps --filter 'fsx'
http://pulpito.ceph.com/trociny-2016-07-28_01:58:45-rbd-wip-mgolub-testing---basic-vps/

teuthology-suite --suite rbd --ceph wip-mgolub-testing --owner mgolub --email mgolub@mirantis.com --machine-type vps --filter 'python_api_tests'
http://pulpito.ceph.com/trociny-2016-07-28_02:00:35-rbd-wip-mgolub-testing---basic-vps/

teuthology-suite --suite rbd --ceph wip-mgolub-testing --owner mgolub --email mgolub@mirantis.com --machine-type vps --filter 'c_api_tests'
http://pulpito.ceph.com/trociny-2016-07-28_02:03:05-rbd-wip-mgolub-testing---basic-vps/
```

Most of the tests have finished, so far the reported failures looks only for valgrind tests due to  "'ceph-fs-common-dbg' was not found".

File: src/rocksdb [hunk 2]
```diff
@@ -1 +1 @@
-Subproject commit 6797e6ffacd90470140f64cea068970bb3a30aaf
+Subproject commit 5647fa427c4eab87c0caf8bcbb91f5846a4623f9
```
Comment (tchaikov) [inline src/rocksdb]: this change is introduced by mistake.

Comment (dillaman) [inline src/rocksdb]: Thanks for the catch
